src/backend/gl/gl_common.c

```diff
@@ -623,13 +624,119 @@ bool gl_kernel_blur(backend_t *base, double opacity, void *ctx, const rect_t *ex
 	return true;
 }
 
-bool gl_dual_kawase_blur(backend_t *base, double opacity attr_unused, void *ctx,
-                         const rect_t *extent attr_unused, const int nrects attr_unused,
-                         const GLuint vao[2] attr_unused) {
-	auto bctx attr_unused = (struct gl_blur_context *)ctx;
-	auto gd attr_unused = (struct gl_data *)base;
+bool gl_dual_kawase_blur(backend_t *base, double opacity, void *ctx, const rect_t *extent,
+                         const int width, const int height, const int nrects,
+                         const GLuint vao[2]) {
+	auto bctx = (struct gl_blur_context *)ctx;
+	auto gd = (struct gl_data *)base;
+
+	int dst_y_screen_coord = gd->height - extent->y2,
+	    dst_y_fb_coord = bctx->fb_height - extent->y2;
+
+	// Reduce number of iterations until the last one renders at least 1px in both
+	// dimensions
+	int iterations = bctx->blur_texture_count;
+	while (((width / (1 << iterations)) < 1 || (height / (1 << iterations)) < 1) &&
+	       iterations > 0) {
+		--iterations;
+	}
+	assert(iterations < (int)(sizeof(int) * 8));
```

Comment:
What does this check for?

Reply:
Trying to explain `clang-tidy` the scale factor won't be too large [over here](https://github.com/yshui/picom/blob/b6987352499bd23b8eae73553e64aa9169bcb6e6/src/backend/gl/gl_common.c#L734). But I am not quite comfortable with how it turned out.

Reply:
You can have a `scale` variable, and double it every iteration. clang-tidy probably won't complain about that.

Reply:
Adding another variable to keep track of the current scaling factor worked out nicely.